### ISSUE TITLE
`HOMEBREW_VERBOSE=1 HOMEBREW_DEBUG=1 brew livecheck --json` prints `bundle clean` to the stdout

### ISSUE THREAD

**Issue by gibfahn** [Homebrew/brew#16134]
### `brew doctor` output

```console
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: The filesystem on /System/Volumes/Data appears to be case-sensitive.
The default macOS filesystem is case-insensitive. Please report any apparent problems.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```console
❯ brew config
HOMEBREW_VERSION: 4.1.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 35746e0a6ba6c3c5cfe56d99f79d9ec9f52ee15f
Last commit: 2 days ago
Core tap JSON: 24 Oct 15:36 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BAT_THEME: TwoDark
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 14.0.3 build 1403
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5.1-x86_64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
```

### What were you trying to do (and why)?

Trying to use livecheck on a formula in a tap. Parsing the JSON output failed because of the extra text to the stdout.

### What happened (include all command output)?

```console
❯ HOMEBREW_VERBOSE=1 HOMEBREW_DEBUG=1 brew livecheck --newer-only --json --formula Formula/foo.rb 2>/dev/null
/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/bin/bundle clean
[
  {
    "formula": "foo",
    ...
  }
]
```

### What did you expect to happen?

Only the JSON to be written to the stdout.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
HOMEBREW_VERBOSE=1 HOMEBREW_DEBUG=1 brew livecheck --json --cask hammerspoon 2>/dev/null
```

**Comment by gibfahn** [comment]
I tried to find where this was being printed but couldn't work it out. 

As a side-note, what would folks think about the default setting for `ohai` and friends to be to print to stdout?

**Comment by MikeMcQuaid** [comment]
> As a side-note, what would folks think about the default setting for `ohai` and friends to be to print to stdout?

Do you mean stderr or stdout? It seems from this issue that you're asking for stdout output to go to stderr.

`ohai` should not go to stderr by default. It's not an error.

This seems extremely specific to "commands/flags that only output JSON". In those cases: outputting the non-JSON output to `stderr` seems reasonable. That said, though: it does feel like with `HOMEBREW_DEBUG=1` you are actively asking for extra, non-error output so perhaps it should just error out instead.

**Comment by gibfahn** [comment]
>Do you mean stderr or stdout? It seems from this issue that you're asking for stdout output to go to stderr.

Sorry, typo, fixed

>ohai should not go to stderr by default. It's not an error.

[Opinions vary](https://stackoverflow.com/questions/4919093/should-i-log-messages-to-stderr-or-stdout), but it's common for all logging to go to stderr, not just error logs. In general if a command returns something I want to use, that goes to `stdout` and any logging goes to `stderr`, that way you can pipe it.

As an example, if there was an `ohai` in `brew list`, even though that's not explicitly designed for machine parsing, it's still something that often gets put into a shell pipeline. As an example:

```shell
if brew list | grep -qx gh; then
  echo >&2 "gh is installed!"
fi
```

This could be broken by extraneous logging going to stdout.

**Comment by gibfahn** [comment]
This is an example that actually logs to stdout even without the debug logs:

https://github.com/Homebrew/brew/blob/64d309c4cf3a199a2c895980abc10b8bd53b9cbc/Library/Homebrew/download_strategy.rb#L442

**Comment by MikeMcQuaid** [comment]
> As an example, if there was an `ohai` in `brew list`, even though that's not explicitly designed for machine parsing, it's still something that often gets put into a shell pipeline.

This has been the case for >14 years at this point so I don't think is worth changing.

Where commands have specific implications of their output and need to output a well-structured document i.e. JSON: we should avoid breaking the output with `HOMEBREW_DEBUG` or `HOMEBREW_VERBOSE`, I agree. I see two paths to that:
1. all debug/verbose logging is handled differently in a "JSON mode" (or similar "porcelain" mode)
2. we just error out if you try to do `HOMEBREW_DEBUG` or `HOMEBREW_VERBOSE`, pass `--json` (or similar) and are also writing to a file/pipe.

> This is an example that actually logs to stdout even without the debug logging configured:

This is not a debug log, it's informational.

**Comment by gibfahn** [comment]
>This has been the case for >14 years at this point so I don't think is worth changing.

The alternative is I file an issue every time one of these gets through, so worth it is really in the eye of the maintainer 😁 . There are currently no default logs 

>all debug/verbose logging is handled differently in a "JSON mode" (or similar "porcelain" mode)

That seems like more work than changing the default of `ohai` and friends, but happy either way.

>we just error out if you try to do HOMEBREW_DEBUG or HOMEBREW_VERBOSE, pass --json (or similar) and are also writing to a file/pipe.

In CI/automation pipelines I want debug and verbose output, I just want it to log to stderr like it does in all the other standard unix tools on my system (e.g. `ls`, `du` etc.).

>This is not a debug log, it's informational.

I wasn't saying it was, I was saying it was to the stdout. Put another way, it's something that can break running some commands even without setting the verbose and debug flags, so would need to be fixed manually. I haven't filed because I can't remember how I hit the error, but it broke automation for us.

**Comment by Bo98** [comment]
Updating https://github.com/Homebrew/brew/blob/64d309c4cf3a199a2c895980abc10b8bd53b9cbc/Library/Homebrew/utils.rb#L44-L50 to respect the `:out` key passed to `options` could be a reasonable generic tweak.

For reference on how this is called:

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/gems.rb#L320

and how the redirection part of `options` works:

```
redirection:
    key:
      FD              : single file descriptor in child process
      [FD, FD, ...]   : multiple file descriptor in child process
    value:
      FD                        : redirect to the file descriptor in parent process
      string                    : redirect to file with open(string, "r" or "w")
      [string]                  : redirect to file with open(string, File::RDONLY)
      [string, open_mode]       : redirect to file with open(string, open_mode, 0644)
      [string, open_mode, perm] : redirect to file with open(string, open_mode, perm)
      [:child, FD]              : redirect to the redirected file descriptor
      :close                    : close the file descriptor in child process
    FD is one of follows
      :in     : the file descriptor 0 which is the standard input
      :out    : the file descriptor 1 which is the standard output
      :err    : the file descriptor 2 which is the standard error
      integer : the file descriptor of specified the integer
      io      : the file descriptor specified as io.fileno
```

**Comment by RandomDSdevel** [comment]
> [Opinions vary](https://stackoverflow.com/questions/4919093/should-i-log-messages-to-stderr-or-stdout), but it's common for all logging to go to stderr, not just error logs. In general if a command returns something I want to use, that goes to `stdout` and any logging goes to `stderr`, that way you can pipe it.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(_Aside:_  Not relevant in Ruby land, but:  This is why C++'s standard library provides [the '`std::clog`' output stream](https://en.cppreference.com/w/cpp/io/clog), though it's technically just another handle to '`std::cerr`'/'`::stderr`,' so that kind of makes the point at least a bit moot.  The difference is that `clog` is buffered by default, while `cerr` defaults to being un-buffered.)  

---

>  1. all debug/verbose logging is handled differently in a "JSON mode" (or similar "porcelain" mode)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_Nods._  Homebrew doesn't have an official porcelain mode like Git does; from what I understand, it hasn't really needed one yet.

**Comment by apainintheneck** [comment]
> all debug/verbose logging is handled differently in a "JSON mode" (or similar "porcelain" mode)

This would be ideal in my opinion but it might require a large number of changes to the codebase since it hasn't been the default for 14 years. I also wonder if commands like `brew list` that have special handling for data that is not flowing to a tty could be updated as well.

I also don't think that the distinction between debug and informational output is necessarily that helpful. If the output is not what is expected from stdout and the only necessary change is to point to stderr instead, we should be open to changing those print statements. That doesn't mean we should go around and change everything but it certainly seems reasonable when someone can point to something specific and say this line of code is breaking my output.

**Comment by gibfahn** [comment]
> > The alternative is I file an issue every time one of these gets through, so worth it is really in the eye of the maintainer 😁 .
> 
> Hopefully it goes without saying but: please don't do this.

Yeah that was meant to be "this is what I don't want to do". But to be clear, if there is info-level logging in something that outputs JSON (or is otherwise reasonably expected to be parsed like `brew list`), I'll still have to file a new issue/PR for every `puts` or `ohai` that causes a regression unless we fix it in one of these two ways.

> In what respect do either of these tools log debug or verbose output to stderr? Neither have equivalent `--verbose` or `--debug` flags.

Fair point, I didn't think about that because I don't care about whether it's verbose or info, I don't want logging going to the same place as the output of the program. First example I can think of then is:

```shell
GIT_TRACE=1 git status >/dev/null
```

Git has plumbing commands, but they know people are still going to (ab)use the porcelain in scripts.

> > I haven't filed because I can't remember how I hit the error, but it broke automation for us.
> 
> This is an issue with your automation rather than with Homebrew, sorry.

Yep totally, I want to use it as an example here of the kind of place I've seen this before, but I'm not asking you to fix it until I can repro the issue.

> > to respect the `:out` key passed to `options` could be a reasonable generic tweak.
> 
> This is a good idea for `--debug`. I'm less convinced for `--verbose`.
>
> > I also think it's reasonable for all print statements that are hidden behind verbose or debug checks to go to stderr.
> 
> I agree with this for debug but not for verbose.

What's the difference here? Which logs are set with debug but not verbose (or vice versa)?

> The possible acceptable solutions I see for this are:
> 
> 1. `HOMEBREW_VERBOSE` or `HOMEBREW_DEBUG` output (but not behaviour, `--debug`'s differing behaviour on e.g. handling some exceptions is likely useful) is silently (or with a `opoo` warning to `stderr) never output when you pass `--json`
> 2. `HOMEBREW_VERBOSE` or `HOMEBREW_DEBUG` env vars/flags conflict with `--json`: you cannot use them both `--debug` and `--json` at once
> 3. `HOMEBREW_DEBUG` (but not `HOMEBREW_VERBOSE`) output is output to `stderr` when you pass `--json`
> 
> I can also see an argument that e.g. longer exception backtraces from `HOMEBREW_DEBUG` should be output to `stderr` (if they are not already).
> 
> We should not turn this issue into "Homebrew's verbose or debug or informational output more generally should always go to `stderr`) because 1) it blows the scope up to be so hilariously large it'll never get done and 2) because it's just a bad idea (for reasons above).

I thought just changing the default output to stderr for logging would be easier than trying to make more targeted changes (like different things for JSON/non-JSON output in every command), because you could "just" change the logging functions. Obviously I don't have much context here though.

**Comment by gibfahn** [comment]
> Updating `https://github.com/Homebrew/brew/blob/64d309c4cf3a199a2c895980abc10b8bd53b9cbc/Library/Homebrew/utils.rb#L44-L50` to respect the `:out` key passed to `options` could be a reasonable generic tweak.

Is this basically saying "if stdout is not a tty write all logs to stderr"? If so that would work great for me if it works for you folks.

**Comment by Bo98** [comment]
> Is this basically saying "if stdout is not a tty write all logs to stderr"? If so that would work great for me if it works for you folks.

More along the lines of: we explicitly tell `bundle clean` to output all logs to stderr, so `system` should respect that choice for the logs it adds itself (i.e. that specific verbose log line). Telling `system` to output to stderr and it doesn't do so completely sounds like unexpected behaviour to me.

`brew livecheck --json` is exactly the reason I made `bundle clean` output to stderr to begin with.

**Comment by gibfahn** [comment]
So to get more specific, as I understand it there are the following:

Library/Homebrew/extend/kernel.rb has:

- `ofail`, `odie`, `opoo`, `onoe` -> warning and fatal errors
- `odebug` -> debug error (but has an `always_display` which I guess makes it an info error)
- `ohai` -> info errors

And there's also:

- `puts`: info error without a header / colouring, used directly in the codebase
- Some subcommands brew executes might output directly to the stdout (e.g. bundler)

The warning/error logs as far as I can tell everyone is happy to write to `stderr`.

The debug stuff can be "fixed" by making `--json` and `HOMEBREW_DEBUG`/`HOMEBREW_VERBOSE` not usable together (although that seems like a regression to me from a debugging my CI PoV).

So the interesting question is the info level logging, as that's only (as far as I can tell) fixable by:
- Going through all output for anything that could be touched by a "reasonably piped to something brew command" to make sure it doesn't log to stderr
- Making all info logging go to stderr

There's also the question of what a non-core/cask formula should do if it wants to have an `ohai`, for example in a custom `livecheck` block. I've been doing `STDERR.ohai` instead of `ohai`, but I don't know if that's expected for all third-party code.

**Comment by MikeMcQuaid** [comment]
> What's the difference here? Which logs are set with debug but not verbose (or vice versa)?

Debug is generally extra output only for Homebrew maintainers/contributors. Verbose is generally extra information that could be useful for users. It varies a bunch from command to command.

> More along the lines of: we explicitly tell `bundle clean` to output all logs to stderr, so `system` should respect that choice for the logs it adds itself (i.e. that specific verbose log line). Telling `system` to output to stderr and it doesn't do so completely sounds like unexpected behaviour to me.

Agreed.

> * `ofail`, `odie`, `opoo`, `onoe` -> warning and fatal errors

Correct.

> * `odebug` -> debug error (but has an `always_display` which I guess makes it an info error)

Not errors, just information.

> * `ohai` -> info errors

Never errors, just prominent headline information that can be sometimes intentionally truncated.


> The warning/error logs as far as I can tell everyone is happy to write to `stderr`.

When they are definitely errors/warnings and a given thing e.g. `odie`/`opoo` is only ever used for error or warning output: yes.

> * Going through all output for anything that could be touched by a "reasonably piped to something brew command" to make sure it doesn't log to stderr

I think even "reasonably piped" here can be initially filtered to "structured data that's actually broken otherwise" i.e. JSON for `brew livecheck --json | jq .` and extended to "unstructured data that's less useful otherwise" e.g. `brew options $(brew list --formula)`

> There's also the question of what a non-core/cask formula should do if it wants to have an `ohai`, for example in a custom `livecheck` block. I've been doing `STDERR.ohai` instead of `ohai`, but I don't know if that's expected for all third-party code.

Formula and casks should never use `ohai`. Arguably they shouldn't print output at all but, if really necessary, it should probably be with `odebug`.

**Comment by gibfahn** [comment]
>Formula and casks should never use ohai. Arguably they shouldn't print output at all but, if really necessary, it should probably be with odebug.

Given the docs say (https://docs.brew.sh/Formula-Cookbook#messaging):

>Three commands are provided for displaying informational messages to the user:
>
>- ohai for general info
>- opoo for warning messages
>- odie for error messages and immediately exiting

I think it's going to be hard to change that at this point.

**Comment by gibfahn** [comment]
>I think even "reasonably piped" here can be initially filtered to "structured data that's actually broken otherwise" i.e. JSON for brew livecheck --json | jq . and extended to "unstructured data that's less useful otherwise" e.g. brew options $(brew list --formula)

Fair enough, and I guess just having an integration test for this for each of the common commands that runs e.g. `brew livecheck --json` and parses the output as JSON should be enough to catch _most_ issues.

**Comment by MikeMcQuaid** [comment]
> I think it's going to be hard to change that at this point.

It doesn't really matter either way for this conversation: this stuff shouldn't be going to `stderr` even as described

**Comment by MikeMcQuaid** [comment]
> Fair enough, and I guess just having an integration test for this for each of the common commands that runs e.g. `brew livecheck --json` and parses the output as JSON should be enough to catch _most_ issues.

Yes, having something like this for `HOMEBREW_DEBUG` and checking the JSON output is parseable would seem like a reasonable fix 👍🏻

**Comment by Bo98** [comment]
The original issue described here should be fixed now in 4.1.19.

**Comment by gibfahn** [comment]
Thanks for the fix, maybe this won't come up again.
